File: mixed/f1/group1/LongestCommonPrefix.java
// Write a function to find the longest common prefix string amongst an array of strings.
// If there is no common prefix, return an empty string "".
// Sample 1:
// Input: strs = ["flower","flow","flight"]
// Output: "fl"

import java.util.Arrays;
class LongestCommonPrefix {
    public static void main(String...args) {
        String[] strs = {"flower","flow","flight"};
        System.out.println("Longest...: " + longestCommonPrefix(strs));
    }

    static String longestCommonPrefix(String[] strs) {
        Arrays.sort(strs);
        String s1 = strs[0];
        String s2 = strs[strs.length - 1];
        int idx = 0;
        while (idx < s1.length() && idx < s2.length()) {
            if (s1.charAt(idx) == s2.charAt(idx)) {
                idx++;
            } else {
                break;
            }
        }
        return s1.substring(0, idx);
    }
}

File: bytopic/array/python/group1/longest_common_prefix.py
"""
Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string "".

Example:
Input: strs = ["flower","flow","flight"]
Output: "fl"

Tag: 20/150
Tag: 14/2927, R13/2936 (overall frequency ranking)
"""


def longest_common_prefix(input_: list[str]) -> str:
    ans = ""
    input_ = sorted(input_)
    first = input_[0]
    last = input_[-1]
    for i in range(min(len(first), len(last))):
        if first[i] != last[i]:
            return ans
        ans += first[i]
    return ans


def main():
    input_ = ["flower", "flow", "flight"]
    print(longest_common_prefix(input_))


if __name__ == "__main__":
    main()
